cemeng/air-api
# Request Rate Limiter

This PR implements Request Rate Limiter functionality.

# Solution Design

## Main files

```RequestLimit``` is the main code that checks whether or not an IP has exceeded its limit.
```RequestCounter``` contains the count of the request from an IP. It also contains the expiry of the limit.
```RequestCounterDataStore``` is the data store storing ```RequestCounter``` entries.

## Data store

```ActiveSupport::Cache::MemoryStore``` is chosen as the data store for the solution, for the following reason:
* It is simple to use, with a well known ```Rails.cache``` API
* It has built in expiring mechanism (by specifying expires_in option)
* It requires no external software to be used other than Rails
* It is fast as it is in memory 

However this data store has limitations:
* This data store will not work if the Rails server used is Phusion Passenger or Puma clustered mode, as it data store stores everything into memory in the same process. In multiple Rails server processes - the server process instances won't be able to share cache data (reference: http://api.rubyonrails.org/classes/ActiveSupport/Cache/MemoryStore.html).
* This data store **will not be persisted** when Rails server stopped. 

If the solution needs to support the above case then perhaps a caching system such as *memcached* or *redis* is more suitable.

## Flexibility / Easy to Change

I tried to balance the possible future requirements with YAGNI (you aint gonna need it) when implementing the solution, as such:
* ```RequestCounterDataStore``` is extracted as its own standalone component (implemented as singleton). If the underlying datastore need changing, only this file needs to be changed.
* ```RequestLimit``` takes ```method``` and ```endpoint``` - which gives it a more granular rate limiting control. For example: for the purpose of this exercise, the 100 limit is given to an IP for **all** endpoints and **all** request method, that can be changed to 100 request limit to an IP only for an endpoint. 

# Development Approach

The following are best practices that I used to develop the solution:
* The solution has been mostly **Test Driven**.  I am actually trying out **Minitest** for this (I'm an RSpec user normally), hence I am sure some parts of the tests can be better written.
* The code has **100% test coverage**. I am using **SimpleCov** to ensure it.
* The code has also adhered to Ruby coding style as suggested by **RuboCop** (I'm strongly recommend using this in a team environment).
* The commits have been written with enough comments. I have left the commits as is, e.g not squashed to show my working process.

# Manual Testing (MacOS)

Ensure the rails server is up:
```
$ rails s
```

Then run the following on the console, this will hit the endpoint 102 times:
```
$ repeat 102 curl http://0.0.0.0:3000/home
```

# Further Work / Possible Improvements

* ```RequestLimit``` and its associated classes can be extracted as a gem and potentially can be used on different projects and even open sourced.
* I would like to introduce  [Flip gem](https://github.com/pda/flip) for feature flagging the rate limiter. This way, we can turn on / off rate limiter easily in production.
* If there is a need for these requests to be audited, the requests should be persisted in a database. The counter mechanism can stay the same.
* The expiry and max allowed request are hardcoded in ```RequestLimit``` class. When they need to change in production, a deployment is required. These values be extracted into a database.
* The expiry and max allowed values can be made to be IP address specific or even endpoint specific. For example: Air company might have different user tiers with different request quota. This would again require extracting them into database.
